include/lsst/afw/image/detail/StorableMap.h

```diff
@@ -68,6 +70,23 @@ class StorableMap final {
     using pointer = value_type*;
     using const_pointer = value_type const*;
 
+private:
+    // Can't map by full Key because we don't have a Key comparator that
+    // respects supertypes, and such a comparator can't be symmetric anyway
+    using _Impl = std::unordered_map<key_type::KeyType, mapped_type>;
+
+    /// Convert a key to _Impl into a key for the StorableMap
+    static value_type _toTypedKey(_Impl::value_type const& v) {
+        return std::make_pair(key_type(v.first), v.second);
+    }
+
+public:
+    using const_iterator = boost::transform_iterator<decltype(&StorableMap::_toTypedKey),
+                                                     StorableMap::_Impl::const_iterator>;
+    using iterator = const_iterator;  // Can't return modifiable reference to value_type
+    using reverse_iterator = std::reverse_iterator<iterator>;
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
```

Comment:
Isn't this latter line redundant with the fact that `iterator` is always `const`?

Reply:
These four types are conventional; all the standard containers have them. The fact that my iterators don't return references (let alone modifiable ones) is the real problem. 😞